File: src/main/java/com/exactpro/th2/act/convertors/ConvertorsRequest.java
package com.exactpro.th2.act.convertors;

import com.exactpro.th2.act.grpc.NewOrderSingle;
import com.exactpro.th2.act.grpc.PlaceMessageRequestTyped;
import com.exactpro.th2.act.grpc.Quote;
import com.exactpro.th2.act.grpc.SecurityListRequest;
import com.exactpro.th2.common.grpc.ConnectionID;
import com.exactpro.th2.common.grpc.Message;
import com.exactpro.th2.common.grpc.MessageID;
import com.exactpro.th2.common.grpc.MessageMetadata;

import java.util.Map;

import static com.exactpro.th2.common.value.ValueUtils.toValue;

public class ConvertorsRequest {

    public Message createMessage(PlaceMessageRequestTyped requestTyped) {
        Message requestMessage;
        switch (requestTyped.getMetadata().getMessageType()) {
            case "NewOrderSingle":
                requestMessage = createNewOrderSingle(requestTyped);
                break;
            case "Quote":
                requestMessage = createQuote(requestTyped);
                break;
            case "SecurityListRequest":
                requestMessage = createSecurityListRequest(requestTyped);
                break;
            default:
                requestMessage = null;
                break;
        }
        return requestMessage;
    }

    private Message.Builder messageBuilder(PlaceMessageRequestTyped requestTyped){
        MessageMetadata metadata = requestTyped.getMetadata();

        return Message.newBuilder()
                .setParentEventId(requestTyped.getParentEventId())
                .setMetadata(MessageMetadata.newBuilder()
                        .setMessageType(metadata.getMessageType())
                        .setId(MessageID.newBuilder()
                                .setConnectionId(ConnectionID.newBuilder()
                                        .setSessionAlias(metadata.getId().getConnectionId().getSessionAlias()))));
    }

    private Message createNewOrderSingle(PlaceMessageRequestTyped requestTyped) {
        NewOrderSingle newOrderSingle = requestTyped.getMessageTyped().getNewOrderSingle();

        return messageBuilder(requestTyped).putAllFields(
                Map.of("Price", toValue(newOrderSingle.getPrice()),
                        "OrderQty", toValue(newOrderSingle.getOrderQty()),
                        "Side", toValue(newOrderSingle.getSide()),
                        "TimeInForce", toValue(newOrderSingle.getTimeInForce()),
                        "UserID", toValue(newOrderSingle.getUserId()),
                        "Instrument", toValue(newOrderSingle.getInstrument())
                )).build();
    }

    private Message createQuote(PlaceMessageRequestTyped requestTyped) {
        Quote quote = requestTyped.getMessageTyped().getQuote();

        Message.Builder noQuoteQualifiers = Message.newBuilder();
        for (Quote.QuoteQualifier quoteQualifier : quote.getNoQuoteQualifiersList()) {
            noQuoteQualifiers.putFields("QuoteQualifier", toValue(quoteQualifier.getQuoteQualifier()));
        }

        return messageBuilder(requestTyped).putAllFields(
                Map.of("NoQuoteQualifiers", toValue(noQuoteQualifiers),
                        "OfferPx", toValue(quote.getOfferPx()),
                        "OfferSize", toValue(quote.getOfferSize()),
                        "QuoteID", toValue(quote.getQuoteId()),
                        "Symbol", toValue(quote.getSymbol()),
                        "SecurityIDSource", toValue(quote.getSecurityIdSource()),
                        "BidSize", toValue(quote.getBidSize()),
                        "BidPx", toValue(quote.getBidPx()),
                        "SecurityID", toValue(quote.getSecurityId()),
                        "QuoteType", toValue(quote.getQuoteType())
                )).build();
    }

    private Message createSecurityListRequest(PlaceMessageRequestTyped requestTyped) {
        SecurityListRequest securityListRequest = requestTyped.getMessageTyped().getSecurityListRequest();

        return messageBuilder(requestTyped).putAllFields(
                Map.of("SecurityListRequestType", toValue(securityListRequest.getSecurityListRequestType()),
                        "SecurityReqID", toValue(securityListRequest.getSecurityReqId())
                )).build();
    }
}

File: src/main/kotlin/com/exactpro/th2/act/convertors/ConvertorsRequest.kt
/*
 * Copyright 2022-2022 Exactpro (Exactpro Systems Limited)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.exactpro.th2.act.convertors

import com.exactpro.th2.act.grpc.NoPartyIDs
import com.exactpro.th2.act.grpc.PlaceMessageRequestTyped
import com.exactpro.th2.common.grpc.ConnectionID
import com.exactpro.th2.common.grpc.Message
import com.exactpro.th2.common.grpc.MessageID
import com.exactpro.th2.common.grpc.MessageMetadata
import com.exactpro.th2.common.value.toValue


class ConvertorsRequest {
    fun createMessage(requestTyped: PlaceMessageRequestTyped): Message {
        val requestMessage: Message = when (val messageType = requestTyped.metadata.messageType) {
            "NewOrderSingle" -> createNewOrderSingle(requestTyped)
            "Quote" -> createQuote(requestTyped)
            "SecurityListRequest" -> createSecurityListRequest(requestTyped)
            else -> throw IllegalArgumentException("Unsupported request message type $messageType")
        }
        return requestMessage
    }

    private fun messageBuilder(requestTyped: PlaceMessageRequestTyped): Message.Builder {
        val metadata = requestTyped.metadata

        return Message.newBuilder()
            .setParentEventId(requestTyped.parentEventId)
            .setMetadata(
                MessageMetadata.newBuilder()
                    .setMessageType(metadata.messageType)
                    .setId(
                        MessageID.newBuilder()
                            .setConnectionId(
                                ConnectionID.newBuilder()
                                    .setSessionAlias(metadata.id.connectionId.sessionAlias)
                            )
                    )
            )
    }

    private fun createNewOrderSingle(requestTyped: PlaceMessageRequestTyped): Message {
        val newOrderSingle = requestTyped.messageTyped.newOrderSingle
        return messageBuilder(requestTyped).putAllFields(

            mutableMapOf(
                "SecurityID" to newOrderSingle.securityId.toValue(),
                "SecurityIDSource" to newOrderSingle.securityIdSource.toValue(),
                "OrdType" to newOrderSingle.ordType.toValue(),
                "AccountType" to newOrderSingle.accountType.toValue(),
                "OrderCapacity" to newOrderSingle.orderCapacity.toValue(),
                "OrderQty" to newOrderSingle.orderQty.toValue(),
                "DisplayQty" to newOrderSingle.displayQty.toValue(),
                "Price" to newOrderSingle.price.toValue(),
                "ClOrdID" to newOrderSingle.clOrdId.toValue(),
                "SecondaryClOrdID" to newOrderSingle.secondaryClOrdId.toValue(),
                "Side" to newOrderSingle.side.toValue(),
                "TimeInForce" to newOrderSingle.timeInForce.toValue(),
                "TransactTime" to newOrderSingle.transactTime.toValue(),
                "TradingParty" to Message.newBuilder().putFields("NoPartyIDs",
                            createNoPartyIdsFields(newOrderSingle.tradingParty.noPartyIdsList).toValue()).toValue(),
                "Symbol" to newOrderSingle.symbol.toValue()
            )
        ).build()
    }

    private fun createQuote(requestTyped: PlaceMessageRequestTyped): Message {
        val quote = requestTyped.messageTyped.quote
        val noQuoteQualifiers = Message.newBuilder()
        for (quoteQualifier in quote.noQuoteQualifiersList) {
            noQuoteQualifiers.putFields("QuoteQualifier", quoteQualifier.quoteQualifier.toValue())
        }
        return messageBuilder(requestTyped).putAllFields(
            mutableMapOf(
                "NoQuoteQualifiers" to noQuoteQualifiers.toValue(),
                "OfferPx" to quote.offerPx.toValue(),
                "OfferSize" to quote.offerSize.toValue(),
                "QuoteID" to quote.quoteId.toValue(),
                "Symbol" to quote.symbol.toValue(),
                "SecurityIDSource" to quote.securityIdSource.toValue(),
                "BidSize" to quote.bidSize.toValue(),
                "BidPx" to quote.bidPx.toValue(),
                "SecurityID" to quote.securityId.toValue(),
                "NoPartyIDs" to createNoPartyIdsFields(quote.noPartyIdsList).toValue(),
                "QuoteType" to quote.quoteType.toValue()
            )
        ).build()
    }

    private fun createSecurityListRequest(requestTyped: PlaceMessageRequestTyped): Message {
        val securityListRequest = requestTyped.messageTyped.securityListRequest
        return messageBuilder(requestTyped).putAllFields(
            mutableMapOf(
                "SecurityListRequestType" to securityListRequest.securityListRequestType.toValue(),
                "SecurityReqID" to securityListRequest.securityReqId.toValue()
            )
        ).build()
    }

    private fun createNoPartyIdsFields(listNoPartyIds: List<NoPartyIDs>): List<Message.Builder> {
        val messages: MutableList<Message.Builder> = ArrayList()
        for (noPartyIds in listNoPartyIds) {
            messages.add(
                Message.newBuilder()
                    .putFields("PartyID", noPartyIds.partyId.toValue())
                    .putFields("PartyIDSource", noPartyIds.partyIdSource.toValue())
                    .putFields("PartyRole", noPartyIds.partyRole.toValue())
            )
        }
        return messages
    }
}